editor_drafts/draft-prorock-spice-cose-sd-cwt.md

```diff
@@ -54,287 +51,375 @@ document are to be interpreted as described in [@!RFC2119].
-    unprotected: {
-      ? disclosures: [* claim-pair] / nil
-      COSE_Countersignature0: bstr
-    },
-    payload : bstr / nil,
-    signature : bstr,
-]
-```
+To verify an SD-CWT, the recipient extracts the protected CWT claims from the payload. 
+These CWT claims contain hash digests of the original claim values combined with unique random salts.
 
-## Data Structures
+By performing these steps, the recipient can cryptographically verify the integrity of the protected claims and verify they have not been tampered with or substituted after issuance by the trusted issuer. 
+The disclosures provide the plaintext claim values for utilization by the recipient.
 
-TBD - Describe common data structures in CDDL
+The algorithm for transorming the CWT Claimset mirrors the algorithm defined in SD-JWT.
+The primary differences are that CBOR maps have replaced JSON Objects, and that CBOR labels (integers) have replaced the strings used by SD-JWT which are `_sd` and `...`.
+This specification uses the term `validated_presented_sd_claims` to refer to the final CBOR map which is produced by subsituting disclosed values which are presented, and removing labels marked for redaction.
+
+As described in SD-JWT, if there remain unused SD-CWT disclosures at the end of this procedure the SD-CWT MUST be considered invalid, as if the siganture had failed to verify.
 
 ## Examples
 
-TBD - Provide examples
+TBD - Provide more examples
+
+### Minimal spanning example
+
+The following example contains claims needed to demonstrate redaction of key-value pairs and array elements.
+
+``` cbor-diag
+/ cose-sign1 / 18([
+  / protected / << {
+    / alg / 1 : -35 / ES384 /
+    / typ / 16 : "application/sd+cwt"
+  } >>,
+  / unprotected / {
+    / sd_claims / TBD : h'82501bb4...6c655f32'
+    / sd_kbt    / TBD : << [
+      / protected / << {
+          / alg / 1 : -35 / ES384 /
+          / typ / 16 : "application/kb+cwt"
+      } >>,
+      / unprotected / {},
+      / payload / <<
+        / cnonce / 39   : h'e0a156bb3f',
+        / aud     / 3   : "https://verifier.example",
+        / iat     / 6   : 1783000000,
+        / sd_hash / TBD : h'c341bb4...a5f3f',
+      >>,
+      / signature / h'1237af2e...6789456'
+    ] >>
+  },
+  / payload / <<
+    / iss / 1   : "https://issuer.example",
+    / sub / 2   : "https://device.example",
+    / exp / 2   : 1883000000,
+    / iat / 2   : 1683000000,
+    / cnf / 8   : {
+      / cose key / 1 : {
+        / alg: ES256 /  3: 35,
+        / kty: EC2   /  1: 2,
+        / crv: P-256 / -1: 1,
+        / x / -2: h'768ed8...8626e',
+        / y / -3: h'6a48cc...fd5d5'
+      }
+    },
+
+    / sd_alg / TBD        : -16, / SHA-256 /
+    / redacted_keys / TBD : [ 
+      / redacted age_over_18 / h'abbd...efef',
+      / redacted age_over_21 / h'132d...75e7',
+    ]
+    / swversion / 271 : [
+      "3.5.5",
+      { / redacted version / TBD: h'45dd...87af }
+    ]
+  >>,
+  / signature / h'3337af2e...66959614'
+])
+```
```

Comment:
I made a few changes. 

- I put the disclosed values in sd_claims of the SD-CWT
- I added sd_alg to the kwt (since it also has a hash
- I replaced on array element TBD with "...". I think this needs to be something unusual like "..." because it is inside an arbitrary array that IETF does not control.
- I added an example of a disclosure inside a deeper level of the SD-CWT hierarchy
- I added a kid to the SD-CWT
- assigned all the TBD assignments to TBD1, TBD2, etc.
- moved some comments about values to the right side of the value

```suggestion
``` cbor-diag
/ cose-sign1 / 18([
  / protected / << {
    / alg / 1  : -35 / ES384 /
    / typ / 16 : "application/sd+cwt"
    / kid /    : 'https://issuer.example/cwt-key3'
  } >>,
  / unprotected / {
    / sd_claims / TBD1 : [  /these are the three disclosures/
        <<[
            /salt/   h'c93c7ff5...72c71e26',
            /claim/  "age_over_18",
            /value/  true
        ]>>,
        <<[
            /salt/   h'399c641e...2aa18c1e',
            /claim/  "region",
            /value/  "ca" /California/
        ]>>,
        <<[
            /salt/   h'82501bb4...6c655f32',
            /value/  "4.1.7"
        ]>>
    ]
    / sd_kbt    / TBD2 : << [
      / protected / << {
          / alg / 1 : -35 / ES384 /
          / typ / 16 : "application/kb+cwt"
      } >>,
      / unprotected / {},
      / payload / <<
        / cnonce / 39    : h'e0a156bb3f',
        / aud     / 3    : "https://verifier.example",
        / iat     / 6    : 1783000000,
        / sd_alg  / TBD4 : -16  /SHA-256/ 
        / sd_hash / TBD3 : h'c341bb...4a5f3f',  /hash of sd_claims   /
                                                /using hash in sd_alg/
      >>,
      / signature / h'1237af2e...6789456'
    ] >>
  },
  / payload / <<
    / iss / 1   : "https://issuer.example",
    / sub / 2   : "https://device.example",
    / exp / 2   : 1883000000,
    / iat / 2   : 1683000000,
    / cnf / 8   : {
      / cose key / 1 : {
        / alg: ES256 /  3: 35,
        / kty: EC2   /  1: 2,
        / crv: P-256 / -1: 1,
        / x / -2: h'768ed8...8626e',
        / y / -3: h'6a48cc...fd5d5'
      }
    },
    / sd_alg / TBD4        : -16, / SHA-256 /
    / redacted_keys / TBD5 : [ 
        h'abbd...efef',  / redacted age_over_18 /
        h'132d...75e7',  / redacted age_over_21 /
    ],
    / swversion / 271 : [
      "3.5.5",
      { "...": h'45dd...87af'  /redacted version element/ }
    ],
    "address": {
        "country" : "us",            / United States /
        /redacted_keys/ TBD5 : [
            h'adb70604...03da225b',  / redacted region /
            h'e04bdfc4...4d3d40bc'   / redacted post_code /
        ]
    }
  >>,
  / signature / h'3337af2e...66959614'
])
```
```

Reply:
There is some subtle challenge depending on where you CBOR encode the claims, because you want the hash to be over the encoded bytes, but you also want to have the disclosures decoded for application processing.

I would tend to encode the whole disclosure set once << [...], [...], ...> ... vs [ <<>>, <<>> ] ... Because the kwt will contain the hash of the stabilized << [...], [...], ...> ... With no risk of array reordering breaking things.

But we can adjust this further in later pull requests.